### PR TITLE
Introduce guardexpr, aot autograd guarding of duplicates into torch._guards

### PR DESCRIPTION
**Pull request by voznesenskym**
cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_guards.py [hunk 1]
```diff
@@ -183,6 +183,34 @@ def set_export_info(self, guard_type, guarded_class, code_list, obj_weakref):
 
 T = TypeVar("T")
 
+"""
+Parent structure for guard env expressions.
+A GuardEnvExpr can have any subtype.
+Note: All subtypes must be handled exhaustively in
+torch._dynamo.guards._parse_guard_env_guards to avoid a RuntimeError.
```
Comment (voznesenskym) [inline torch/_guards.py:190]: note: this note might be stale...

File: torch/testing/_internal/common_utils.py [hunk 2]
```diff
@@ -4052,6 +4053,22 @@ def custom_op(opname, symbolic_fn, opset_version):
...
 
+def outs_and_grads(fn, graph_inps, inps):
+    outs = fn(*graph_inps)
+    for out in pytree.tree_flatten(outs)[0]:
+        if isinstance(out, torch.Tensor) and out.requires_grad:
+            out.sum().backward(retain_graph=True)
+    grads = [inp.grad for inp in pytree.tree_flatten(inps)[0]]
+    for inp in pytree.tree_flatten(inps)[0]:
+        inp.grad = None
+    return outs, grads
```
Comment (ezyang) [inline torch/testing/_internal/common_utils.py:4064]: I swear we already have a copy of this helper somewhere :think:

File: torch/_guards.py (hunk 1 above)
Comment (ezyang) [reply torch/_guards.py:190]: if it's stale can we update it?

File: torch/_dynamo/guards.py [hunk 3]
```diff
@@ -506,6 +512,27 @@ def compile_check_fn(
             )
             verbose_code_parts.append(f"___check_tensors_verbose({verbose_args})")
 
+        graphargs = self.output_graph.graphargs
```
Comment (ezyang) [inline torch/_dynamo/guards.py:515]: Use `tracked_fakes`, this will also account for duplicate parameters

Comment (ezyang) [reply torch/_dynamo/guards.py:515]: Oh... you're doing it with index positions... hmm......

File: torch/_dynamo/guards.py [hunk 4]
```diff
@@ -506,6 +512,27 @@ def compile_check_fn(
...
+        ] = self.output_graph.tracing_context.guards_context.aotautograd_guards
+        for guard in aotautograd_guards:
+            if isinstance(guard, DuplicateInputs):
+                pos_a = guard.input_pos_a
+                pos_b = guard.input_pos_b
+                assert pos_b < len(graphargs) and pos_a < len(
+                    graphargs
+                ), "Deduped args out of bounds"
+                assert graphargs[pos_a].is_tensor, "Deduped arg must be a tensor"
+                assert graphargs[pos_b].is_tensor, "Deduped arg must be a tensor"
```
Comment (ezyang) [inline torch/_dynamo/guards.py:528]: I'm not hot on using index positions. The strongest argument against it, I think, is that we need to account for duplicates in parameters too, but this is unrepresentable in the current formulation.

To use not-indexes, we have to propagate source names for the args as well as the parameters/buffers. This is annoying, but doable, I think (maybe we store it sideband in AOTConfig.)

**Review by ezyang — APPROVED** [review]
see comments about indexes

File: torch/_dynamo/guards.py (hunk 4 above)
Comment (voznesenskym) [reply torch/_dynamo/guards.py:528]: Discussed offline. Will do a subsequent PR where we move names to here.

**Pull request closed without merging**